proj/hps_accel/src/blocks.cc

```diff
@@ -161,4 +158,73 @@ int32_t MultiplyByQuantizedMultiplier_03(int32_t x,
   return quotient + rounding;
 }
 
+int32_t MultiplyByQuantizedMultiplier_04(int32_t x, int32_t multiplier,
+                                         int shift) {
+  // Saturating Rounding Double High Mul
+  bool positive = x > 0;
+  int64_t a_64(positive ? x : -x);
+  int64_t b_64(multiplier);
+  int64_t ab_64 = a_64 * b_64;
+  if (positive) {
+    ab_64 += (1 << 30);
+  } else {
+    ab_64 += (1 << 30) - 1;
+  }
+  int64_t t = ab_64 >> 31;
+  int32_t product = static_cast<int32_t>(t);
+  if (!positive) {
+    product = -product;
+  }
+
+  // RoundingDivideByPOT implementation
+  int32_t quotient;
+  int32_t mask;
+  switch (shift) {
+    case -3:
+      quotient = product >> 3;
+      mask = (1 << 3) - 1;
+      break;
+    case -4:
+      quotient = product >> 4;
+      mask = (1 << 4) - 1;
+      break;
+    case -5:
+      quotient = product >> 5;
+      mask = (1 << 5) - 1;
+      break;
+    case -6:
+      quotient = product >> 6;
+      mask = (1 << 6) - 1;
+      break;
+    case -7:
+      quotient = product >> 7;
+      mask = (1 << 7) - 1;
+      break;
+    case -8:
+      quotient = product >> 8;
+      mask = (1 << 8) - 1;
+      break;
+    case -9:
+      quotient = product >> 9;
+      mask = (1 << 9) - 1;
+      break;
+    case -10:
+      quotient = product >> 10;
+      mask = (1 << 10) - 1;
+      break;
+    case -11:
+      quotient = product >> 11;
+      mask = (1 << 11) - 1;
+      break;
+    default:
+      quotient = 0;
+      mask = 0;
+      printf("***BAD SHIFT\n");
```

Comment:
Consider also DCHECK'ing the sanity of this shift value (and other expectations) instead of, or in addition to, this printf.

Currently a DCHECK failure just hangs the whole CPU but I was intending at some point to make our assert() and abort() implementation friendlier and print some good messages before it dies.

I'm regularly running through the model without NDEBUG (meaning, do the assertions) to catch mistakes.

Reply:
Done. Thank you for the suggestion. I guess we should also do DCHECK checking in the CI.